### ISSUE TITLE
ARM servers

### ISSUE THREAD

**Issue by financehoncho** [ripple/rippled#1270]
Does rippled compile on ARM? I get 

```
src/beast/beast/hash/xxhasher.h:41:5: error: static assertion failed: 
     static_assert(sizeof(std::size_t)==8, "");
```

in `app_ledger.cpp`.

**Comment by vinniefalco** [comment]
Are you building with a 64 bit architecture or a 32 bit architecture?

**Comment by financehoncho** [comment]
32bit / armv7l

**Comment by vinniefalco** [comment]
rippled requires a 64-bit arch

**Comment by financehoncho** [comment]
oops sorry thanks

**Comment by offlinehacker** [comment]
Too bad, we had this idea too

**Comment by financehoncho** [comment]
arm64 servers coming soon, e.g. https://www.runabove.com/armcloud.xml

**Comment by financehoncho** [comment]
So I have arm64 now, rippled 0.30.0 fails to compile with an error in rocksdb.cpp, line 51:
``` #error Please implement AtomicPointer for this platform.```
I'd be happy to skip RocksDB and use another DB if this helps.

**Comment by financehoncho** [comment]
So I tried to fix this by adding, in ```src/rocksdb2/port/atomic_pointer.h```, underneath:
```
#elif defined(__ARMEL__)
#define ARCH_CPU_ARM_FAMILY 1
```
the following:
```
#elif defined(__AARCH64EL__)
#define ARCH_CPU_ARM_FAMILY 1
```

Also, not sure whether this is necessary, in ```src/protobuf/src/google/protobuf/stubs/platform_macros.h```, underneath
```
#elif defined(__ARMEL__)
#define GOOGLE_PROTOBUF_ARCH_ARM 1
#define GOOGLE_PROTOBUF_ARCH_32_BIT 1
```
I added
```
#elif defined(__AARCH64EL__)
#define GOOGLE_PROTOBUF_ARCH_ARM 1
#define GOOGLE_PROTOBUF_ARCH_64_BIT 1
```

It compiles now but doesn't start, I just get
```
Launching child 1Loading: "/home/admin/.config/ripple/rippled.cfg"
Launching child 2Loading: "/home/admin/.config/ripple/rippled.cfg"
Launching child 3Loading: "/home/admin/.config/ripple/rippled.cfg"
....
```
and so on.

No log being written either.

**Comment by JoelKatz** [comment]
It's crashing or terminating abnormally. Try starting it with `--fg -v`. If you have `gdb`, running it under `gdb` may help. Start gdb with `gdb /path/to/rippled`. Then `run --fg -v`.

**Comment by financehoncho** [comment]
```
Program received signal SIGSEGV, Segmentation fault.
0x00000000ffff0fa0 in ?? ()
```

Sorry I can't be of much help here, not a system level programmer here.

**Comment by JoelKatz** [comment]
It's crashing very early in its startup. It may help to use a debug build.

**Comment by nbougalis** [comment]
I have a 64-bit ARM device which I can test this on. I likely can't get to it for a few days though.

**Comment by financehoncho** [comment]
I'll gladly make my server available to anyone for debugging purposes.

**Comment by tuloski** [comment]
So no way on a 32-bit arch?
Do I give up right now before trying?

**Comment by vinniefalco** [comment]
I would give up on a 32-bit build. We made the decision a couple of years ago to only support 64-bit architectures and this is reflected by some of the design decisions which make going to 32-bit impossible. And I think the memory requirements for rippled can easily exceed 4GB in certain common configurations. While we would like rippled to run on as wide a variety of platforms as possible, I don't relish the hassle of supporting what is essentially two different apps (32-bit and 64-bit).

**Comment by financehoncho** [comment]
Just to clarify this issue is about ARM (64 bit) not 32 bit in any shape. ARM is the future of server hosting I'm sure if even I concede it's low priority for Ripple a.t.m.

**Comment by vinniefalco** [comment]
There's a much better chance of rippled being supported on 64-bit ARM than 32-bit anything

**Comment by tuloski** [comment]
It's a very understandable choice...32-bit is going to be discontinued everywhere.
Fortunately I read this before starting ;)
I'll wait the ODROID with Exynos 8 Octa 8890 64-bit arm :)

**Comment by zsmith928** [comment]
@vinniefalco @tuloski @financehoncho would you guys be interested in free access to on-demand ARM64 infrastructure to test this?  At Packet, we now have a 96 core (2 x 48 Core ThunderX) system as a deployable config that could be useful in testing this build.

**Comment by vinniefalco** [comment]
Just my first impression but there is a considerable cost to maintaining additional platforms, especially a platform for which no developer has hardware in their office. With respect to ARM64, I'm not seeing a correspondingly large gain as an offset but I could be wrong - anyone?

**Comment by vinniefalco** [comment]
I'm closing this for now. We don't regularly build on ARM64 so we have no way of maintaining the quality of our software for that platform.
